File: apps/admin/lib/core/contratos/texto.dart
import 'package:diacritic/diacritic.dart';

/// El espejo en Dart de `packages/contratos/src/texto.ts`. ARQUITECTURA §7:
/// la normalizacion vive en UN lugar, y el panel no tiene la suya.
///
/// **Lo que hace que esto no sea una segunda implementacion disfrazada** es
/// `test/core/contratos/texto_test.dart`, que corre cada fixture de
/// `packages/contratos/generated/contratos.json` contra estas funciones. Las
/// fixtures las calcula el TypeScript, y `scripts/ci/auditar_estados.mjs`
/// verifica que no envejezcan. Sin esa pieza el test compara contra un
/// contrato que ya no existe y pasa por nada.
///
/// **Por que hace falta `diacritic`:** Dart no trae normalizacion Unicode en
/// el SDK —no hay `String.normalize`—, asi que el `NFD` + sacar las marcas
/// combinantes del lado de TypeScript no se puede escribir igual aca. La
/// alternativa era una tabla de caracteres a mano, y una tabla a mano falla
/// **en silencio** con el caracter que nadie penso: `Château` quedaria
/// `ch-teau` y el operador veria ese slug sin entender por que.
///
/// ⚠️ **Donde termina el acuerdo, dicho a proposito.** `removeDiacritics`
/// traduce las ligaduras a ASCII —`Æ` da `ae`, `ß` da `s`— y el `NFD` de
/// TypeScript no: alla `æ` sobrevive a `normalizar` y lo tira el filtro de
/// `clave`. **El contrato son las fixtures**, no la igualdad para toda
/// entrada posible, y ninguna fixture usa una ligadura. Hoy eso no tiene
/// consumidor: el unico que deriva el slug de una bodega es este panel. El
/// dia que TypeScript tambien lo derive, la ligadura entra como fixture y
/// esto se decide.
///
/// El test fija las dos —`ae` y `s`— porque son lo que se MIDIO, no lo que se
/// supuso: la primera version de ese test decia `ss` para `ß` y fallo.

final _espacios = RegExp(r'\s+');
final _noAlfanumerico = RegExp(r'[^a-z0-9]+');
final _guionesDeLosBordes = RegExp(r'^-+|-+$');

/// Minusculas, sin acentos, sin espacios en los extremos y con los internos
/// colapsados en uno. **Conserva los espacios**: alimenta un `contains`.
String normalizar(String texto) =>
    removeDiacritics(texto).toLowerCase().trim().replaceAll(_espacios, ' ');

/// [normalizar] y ademas sin nada que no sea letra latina o numero, para
/// comparar nombres enteros. Puede dar la cadena vacia.
String clave(String texto) => normalizar(texto).replaceAll(_noAlfanumerico, '');

/// El slug que sale de un nombre, o `''` si del nombre no sale ninguno.
///
/// `''` NO cumple `esSlug` de `firestore.rules`, que es exactamente lo que
/// tiene que pasar: quien llama lo lee como "este nombre no sirve" y no
/// guarda. Devolver vacio en vez de lanzar es a proposito — el panel lo
/// muestra mientras se escribe, y un campo a medio llenar no es un error.
String aSlug(String texto) => normalizar(
  texto,
).replaceAll(_noAlfanumerico, '-').replaceAll(_guionesDeLosBordes, '');

/// `true` si la clave de uno contiene a la del otro: "Catena" contra
/// "Catena Zapata" (HU-02.2).
///
/// ⚠️ **Dos claves vacias, o una, dan `false`.** La cadena vacia esta
/// contenida en todas: sin este corte, toda bodega nueva pareceria un
/// duplicado de todo.
bool seParecen(String a, String b) {
  final x = clave(a);
  final y = clave(b);
  if (x.isEmpty || y.isEmpty) return false;
  return x.contains(y) || y.contains(x);
}

File: packages/contratos/src/texto.ts
/**
 * La normalizacion de texto. ARQUITECTURA §7: *"la normalizacion (acentos,
 * mayusculas) vive en `packages/contratos` con sus fixtures. Un solo lugar."*
 *
 * Hasta hoy no vivia en ninguno: lo unico parecido era un slugify escrito a
 * mano adentro de `apps/tienda/src/features/catalogo/PanelDeFiltros.tsx`.
 *
 * Son CUATRO funciones y no una porque BUSCAR Y COMPARAR NO SON LO MISMO:
 *
 *   · `normalizar` conserva los espacios, porque alimenta un `contains`.
 *   · `clave` los saca, porque compara nombres ENTEROS.
 *
 * Si `normalizar` sacara los espacios, buscar "catena zapata" no encontraria
 * nada: el operador escribe con espacios y el texto indexado no los tendria.
 *
 * El panel en Flutter lo ESPEJA en Dart y verifica el espejo contra
 * FIXTURES_DE_TEXTO, que viajan en `generated/contratos.json`. No viaja la
 * implementacion -no se puede transportar en JSON-: viajan pares
 * entrada->salida calculados por el TypeScript de HOY, y
 * `scripts/ci/auditar_estados.mjs` verifica que ese JSON no envejezca.
 */

/**
 * Minusculas, sin acentos, sin espacios en los extremos y con los internos
 * colapsados en uno. Lo que se compara con `contains` al buscar.
 *
 * Saca las marcas combinantes DESPUES de descomponer en NFD: asi `Ñ` es `n` y
 * `Torrontés` es `torrontes` sin una tabla de caracteres escrita a mano.
 */
export function normalizar(texto: string): string {
  return texto
    .normalize('NFD')
    .replace(/\p{M}/gu, '')
    .toLowerCase()
    .trim()
    .replace(/\s+/g, ' ');
}

/**
 * `normalizar` y ademas sin nada que no sea letra latina o numero, para que
 * "Catena Zapata", "catena-zapata" y "CATENA  ZAPATA" den lo mismo.
 *
 * Puede dar la cadena vacia: un nombre hecho solo de puntuacion no tiene
 * clave. Quien la usa TIENE que tratar ese caso -- ver `seParecen`.
 */
export function clave(texto: string): string {
  return normalizar(texto).replace(/[^a-z0-9]+/g, '');
}

/**
 * El slug que sale de un nombre. Cumple `esSlug` de `firestore.rules`
 * -`[a-z0-9]+(-[a-z0-9]+)*`- o devuelve la cadena vacia, que NO lo cumple y
 * que quien llama tiene que leer como "este nombre no sirve".
 *
 * Devolver `''` en vez de lanzar es a proposito: el panel lo muestra mientras
 * se escribe, y una excepcion por cada tecla intermedia no es un error, es el
 * estado normal de un campo a medio llenar.
 */
export function aSlug(texto: string): string {
  return normalizar(texto)
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '');
}

/**
 * true si una clave contiene a la otra. Resuelve el caso que pide HU-02.2:
 * "Catena" contra "Catena Zapata".
 *
 * NO es difusa a proposito. ARQUITECTURA §7: con menos de 200 productos el
 * panel *"no necesita busqueda difusa, asi que no hay una segunda
 * implementacion que mantener sincronizada"*. Un Levenshtein aca seria
 * exactamente esa segunda implementacion.
 *
 * ⚠️ LA CADENA VACIA ESTA CONTENIDA EN TODAS. Sin el corte de abajo,
 * `seParecen('···', 'Norton')` daria true por algebra y TODA bodega nueva
 * pareceria un duplicado de todo.
 */
export function seParecen(a: string, b: string): boolean {
  const x = clave(a);
  const y = clave(b);
  if (x === '' || y === '') return false;
  return x.includes(y) || y.includes(x);
}

// --------------------------------------------------------------- fixtures

/**
 * Las ENTRADAS de las fixtures. Las salidas las calcula `scripts/generar.mjs`
 * con las funciones de arriba, asi que el JSON siempre dice lo que el
 * TypeScript hace hoy -- no lo que alguien escribio que hacia.
 *
 * Cada entrada esta por un motivo:
 */
export const ENTRADAS_DE_TEXTO: readonly string[] = [
  // El caso plano: no tiene nada raro y tiene que salir intacto.
  'Norton',
  // Espacios en los extremos y dobles adentro: los tres los colapsa.
  '  Bodega   Catena ZAPATA ',
  // La tilde del castellano.
  'Torrontés',
  // La enie: el caracter que una tabla escrita a mano olvida.
  'Bodega Ñandú',
  // Ampersand y punto: no son letras, y cada funcion los trata distinto.
  'Ñandú & Cía.',
  // El circunflejo del frances: NFD lo descompone, una tabla a mano no.
  'Château Vieux',
  // La dieresis del castellano.
  'Pingüino',
  // Solo puntuacion: `aSlug` da '' y `seParecen` tiene que dar false.
  '···',
  // Ya viene en forma de slug: normalizar no lo tiene que romper.
  'catena-zapata',
  // Vacio.
  '',
];

/** Los pares que prueban `seParecen`, con el porque de cada uno. */
export const PARES_DE_TEXTO: readonly (readonly [string, string])[] = [
  // El caso de la historia: uno contenido en el otro.
  ['Catena', 'Catena Zapata'],
  // Dos bodegas que no tienen nada que ver.
  ['Catena', 'Norton'],
  // El mismo nombre escrito de dos formas.
  ['Catena Zapata', 'catena-zapata'],
  // La clave vacia contra un nombre real: NO se parecen.
  ['···', 'Norton'],
  // Dos claves vacias.
  ['···', ''],
  // Contencion al reves: el largo primero.
  ['Bodega Ñandú', 'ñandu'],
];
